docs/design-v2.md

```diff
@@ -0,0 +1,132 @@
+instances.
+
+#### `AzDriverNode` Resource and Controller
+
+The `AzDriverNode` custom resource represents a node in the cluster where the Mandalay node 
+plug-in runs. An instance of `AzDriverNode` is created when the node plug-in starts. The node 
+plug-in periodically updates the heartbeat in the `Status` field.
+
+The controller for `AzDriverNode` runs in the controller plug-in. It is responsible for 
+deleting `AzDriverNode` instances that no longer have corresponding nodes in the cluster.
+
+#### `AzVolume` Resource and Controller
+
+The `AzVolume` custom resource represents the managed disk of a `PersistentVolume`. The 
+controller for `AzVolume` runs in the controller plug-in. It watches for and reconciles 
+changes in the `AzVolume` instances.
+
+An `AzVolume` instance is created by the `CreateVolume` API to the CSI Controller plug-in. The 
+`AzVolume` controller responds to the new instance by creating a managed disk using the 
+parameters in the referenced `StorageClass`. When disk creation is complete, the controller 
+sets the `.Status.State` field to `Created` or suitable error state on failure. It also adds a 
+finalizer if creation was successful. The `CreateVolume` request completes when the `AzVolume` 
+state is updated.
+
+To delete a managed disk, the `DeleteVolume` API in the CSI Controller plug-in deletes the 
+corresponding `AzVolume` instance. The controller responds to deletion by garbage collecting 
+the managed disk. When the managed disk has been deleted, the controller removes the finalizer
+from the `AzVolume` instance. The `DeleteVolume` request completes once the `AzVolume` instance
+has been removed from the object store.
+
+#### `AzVolumeAttachment` Resource and Controller
+
+The `AzVolumeAttachment` custom resource represents the attachment of a managed disk to a 
+specific node. The controller for this custom resource runs in the controller plug-in and 
+watches for changes in the `AzVolumeAttachment` instances.
+
+An `AzVolumeAttachment` instance representing the primary node attachment is created by the 
+`ControllerPublishVolume` API in the CSI Controller plug-in. If an instance for the current 
+node already exists, it is updated to represent the primary node attachment. The 
+`AzVolumeAttachment` controller then attaches the manage disk to the primary node (downgrading 
+to mount replica or detaching any previous primary attachment if one exists). It then creates 
+additional `AzVolumeAttachment` instances representing mount replicas and attaches the shared 
+manage disk to a number of backup nodes as specified by the `maxMountReplicaCount` parameter 
+in the `StorageClass` instance of the `PersistentVolumeClaim` of the managed disk. The 
+controller also sets the `.Status.State` field to `Attached` and a finalizer to each 
+`AzVolumeAttachment` instance as each attachment completes. The `ControllerPublishVolume` 
+request is complete when the state of the primary node attachment is complete.
```

Comment:
We do not plan to wait for AzVolumeAttachment to finally go to attached state. Instead, plan is to synchronize at the time of NodeStageVolume, there we should wait for AzVolumeAttachment to succeed before going for staging.
This will allow us extra time to attach after ControllerPublishVolume and before NodeStageVolume is called, and k8s CO will not be blocked in placing the pods.

Reply:
I updated the document to capture this approach.